src/bors/context.rs

```diff
@@ -1,13 +1,32 @@
-use crate::{bors::command::CommandParser, database::DbClient};
-use std::sync::Arc;
+use arc_swap::ArcSwap;
 
-pub struct BorsContext {
+use crate::{bors::command::CommandParser, database::DbClient, github::GithubRepoName};
+use std::{collections::HashMap, sync::Arc};
+
+use super::{GlobalClient, RepositoryClient, RepositoryState};
+
+pub struct BorsContext<Client: RepositoryClient> {
     pub parser: CommandParser,
     pub db: Arc<dyn DbClient>,
+    pub global_client: Arc<dyn GlobalClient<Client>>,
+    pub repositories: Arc<ArcSwap<HashMap<GithubRepoName, Arc<RepositoryState<Client>>>>>,
```

Comment:
- Let's use `GithubRepositoryMap` (and move it into this file).
- The outer `Arc` looks unnecessary (the main program compiles without it).

But I think that here I would do a step back and think about how exactly we want to read from, and write to, the repository state.
What we want to do:
- Mutably add (even possibly remove) new repositories in refresh, if someone adds a new repo to the bot while it is running (I think that we should support this use-case, even though it is quite rare). It's probably better to just modify the map than to replace it altogether, so I would use a `RwLock` around the `HashMap`.
- We need to access a given repository state in per-repo event handling, and also in global event handling, but for that it should be enough that we have `Arc<BorsContext>`. We don't ever need to change the name of a repo, nor its client. Therefore name and client in repository state could be read-only. We do need to atomically replace the config and the permissions, so these two fields inside the state should be behind `ArcSwap`, and they already are, so that's good.

Apart from that, I don't see other use-cases for accessing the state. So I *think* that we can simplify this type just to `RwLock<HashMap<GithubRepoName, RepositoryState<Client>>>`, and that should work. Could you please try it?

Reply:
I see. I was forced to mindlessly spam `Arc<RepositoryState<Client>>` by the trait method `get_repo_state`, cause there was no way to get a reference into a repo inside the hashmap. Now that the trait is gone, I should have remove all those unnecessary Arc.

About RwLock, do you think using RwLock is better than ArcSwap, at least for the hashmap?

Reply:
It's a bit more annoying because of the unwrap, but I think it makes more sense herr. We don't need to replace the whole hashmap during refresh. In fact, we don't want to, because if e.g. the refresh of a single repo fails, we should still keep its old data, and that would be annoying if we'd always need to rebuild the whole map. Also we don't want to replace the repo clients with cached GH tokens.

Btw, while thinking about this more, I realized that it might be more practical to put the repo state into an Arc after all, just to avoid holding on to the hashmap RwLock for the whole duration of a repo event handling (to avoid repo events from starving global events and vice-versa). But that should be an easy change that we can do later, so let's start with the easier approach without the Arc.

Reply:
I see. Since it's only just a matter of reverting a commit, I have added the Arc back to the repo state